parser_library/src/context/hlasm_context.cpp

```diff
@@ -101,253 +105,267 @@ class sysstmt_macro_param_data : public macro_param_data_single_dynamic
 {
-    if (scope.is_in_macro())
+    struct dynamic_data final : public macro_param_data_single_dynamic
     {
-        {
-            auto sect_name = ord_ctx.current_section() ? ord_ctx.current_section()->name : id_index();
+        Func f;
 
-            scope.system_variables.insert(create_system_variable(id_index("SYSECT"), sect_name.to_string(), false));
-        }
+        C_t get_dynamic_value() const override { return f(); }
 
-        {
-            std::string value = left_pad(std::to_string(SYSNDX_), 4, '0');
+        explicit dynamic_data(Func&& f)
+            : f(std::move(f))
+        {}
+    };
 
-            scope.system_variables.insert(create_system_variable(id_index("SYSNDX"), std::move(value), false));
-        }
+    return std::make_unique<dynamic_data>(std::move(f));
+}
 
-        {
-            std::string value = "";
+auto time_sysvars(utils::timestamp now)
+{
+    struct
+    {
+        std::string datc_val;
+        std::string date_val;
+        std::string systime;
+    } result;
+    auto& [datc_val, date_val, systime] = result;
+    auto year = std::to_string(now.year());
+    auto month = std::to_string(now.month());
+    auto day = std::to_string(now.day());
+    if (now.month() < 10)
+        month.insert(0, 1, '0');
+    if (now.day() < 10)
+        day.insert(0, 1, '0');
+
+    datc_val.reserve(8);
+    date_val.reserve(8);
+
+    datc_val.append(year);
+    datc_val.append(month);
+    datc_val.append(day);
+
+    date_val.append(month);
+    date_val.push_back('/');
+    date_val.append(day);
+    date_val.push_back('/');
+    date_val.append(year.c_str() + 2);
+
+    if (now.hour() < 10)
+        systime.push_back('0');
+    systime.append(std::to_string(now.hour()));
+    systime.push_back(':');
+    if (now.minute() < 10)
+        systime.push_back('0');
+    systime.append(std::to_string(now.minute()));
 
-            if (ord_ctx.current_section())
-            {
-                switch (ord_ctx.current_section()->kind)
-                {
-                    case context::section_kind::COMMON:
-                        value = "COM";
-                        break;
-                    case context::section_kind::DUMMY:
-                        value = "DSECT";
-                        break;
-                    case context::section_kind::READONLY:
-                        value = "RSECT";
-                        break;
-                    case context::section_kind::EXECUTABLE:
-                        value = "CSECT";
-                        break;
-                    default:
-                        break;
-                }
-            }
+    return result;
+}
 
-            scope.system_variables.insert(create_system_variable(id_index("SYSSTYP"), std::move(value), false));
-        }
+} // namespace
 
-        {
-            std::string location_counter_name = "";
+std::vector<macro_data_ptr>& hlasm_context::ensure_dynamic_ptrs_count()
+{
+    auto& dp = dynamic_ptrs_vector;
```

Comment:
Consider having an assert to check the sizes
```suggestion
    assert(dp.size() <= scope_stack_.size());
    auto& dp = dynamic_ptrs_vector;
```

Reply:
That condition can be violated... Actually, it should be violated most of the time.

Reply:
Should there at least be an assert for `i < scope_stack_.size()` in the `else` branch or is it always guaranteed to be true?

Reply:
The rest of the code should ensure that condition, but assert for it does make sense.